File: Modules/ServerTypes/LeaveApplication/LeaveApplicationForm.ts
import { StringEditor, LookupEditor, MultipleImageUploadEditor, DateEditor, EnumEditor, BooleanEditor, DecimalEditor, PrefixedContext } from "@serenity-is/corelib";
import { LeaveTypes } from "./LeaveTypes";
import { initFormType } from "@serenity-is/corelib/q";

export interface LeaveApplicationForm {
    EmployeeUpdatedName: StringEditor;
    SuperiorRejectReason: StringEditor;
    HrUpdatedName: StringEditor;
    HrRejectReason: StringEditor;
    EmployeeUpdated: LookupEditor;
    HrUpdated: LookupEditor;
    EmployeeRowId: LookupEditor;
    EmployeeName: StringEditor;
    LeaveReasonId: LookupEditor;
    LeaveDescriptionID: LookupEditor;
    LeaveDescriptions: StringEditor;
    SupportingDocument: MultipleImageUploadEditor;
    StartDate: DateEditor;
    EndDate: DateEditor;
    HalfDay: EnumEditor;
    MorningSession: BooleanEditor;
    AfternoonSession: BooleanEditor;
    EligibleDay: DecimalEditor;
    BalanceLeave: DecimalEditor;
    LeaveTaken: DecimalEditor;
}

export class LeaveApplicationForm extends PrefixedContext {
    static formKey = 'LeaveApplication.LeaveApplication';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!LeaveApplicationForm.init)  {
            LeaveApplicationForm.init = true;

            var w0 = StringEditor;
            var w1 = LookupEditor;
            var w2 = MultipleImageUploadEditor;
            var w3 = DateEditor;
            var w4 = EnumEditor;
            var w5 = BooleanEditor;
            var w6 = DecimalEditor;

            initFormType(LeaveApplicationForm, [
                'EmployeeUpdatedName', w0,
                'SuperiorRejectReason', w0,
                'HrUpdatedName', w0,
                'HrRejectReason', w0,
                'EmployeeUpdated', w1,
                'HrUpdated', w1,
                'EmployeeRowId', w1,
                'EmployeeName', w0,
                'LeaveReasonId', w1,
                'LeaveDescriptionID', w1,
                'LeaveDescriptions', w0,
                'SupportingDocument', w2,
                'StartDate', w3,
                'EndDate', w3,
                'HalfDay', w4,
                'MorningSession', w5,
                'AfternoonSession', w5,
                'EligibleDay', w6,
                'BalanceLeave', w6,
                'LeaveTaken', w6
            ]);
        }
    }
}

[LeaveTypes]; // referenced types

File: Modules/HumanResource/LeaveApplication/LeaveApplication/LeaveApplicationForm.cs
using Serenity.ComponentModel;
using System;

namespace HRMSoftware.LeaveApplication.Forms;

[FormScript("LeaveApplication.LeaveApplication")]
[BasedOnRow(typeof(LeaveApplicationRow), CheckNames = true)]
public class LeaveApplicationForm
{
    //public string RejectedEmployeeName { get; set; }

    // public string ApproveEmployeeName { get; set; }
    public string EmployeeUpdatedName { get; set; }
    [HideOnUpdate, HideOnInsert]
    public string SuperiorRejectReason { get; set; }
    public string HrUpdatedName { get; set; }
    [HideOnUpdate, HideOnInsert]
    public string HrRejectReason { get; set; }
    public int EmployeeUpdated { get; set; }
    public int HrUpdated { get; set; }

    [HalfWidth]
    public int EmployeeRowId { get; set; }
    [HalfWidth]
    public string EmployeeName { get; set; }

    public int LeaveReasonId { get; set; }

    public int LeaveDescriptionID { get; set; }

    public string LeaveDescriptions { get; set; }


    public string SupportingDocument { get; set; }
    [HalfWidth]
    public DateTime StartDate { get; set; }
    [HalfWidth]
    public DateTime EndDate { get; set; }
    public int HalfDay { get; set; }
    [OneThirdWidth]
    public int MorningSession { get; set; }
    [OneThirdWidth]
    public int AfternoonSession { get; set; }
    [HalfWidth]
    public float EligibleDay { get; set; }
    [HalfWidth]
    public float BalanceLeave { get; set; }
    [HalfWidth]
    public float LeaveTaken { get; set; }
}